### PR TITLE
Allow EXPOSE <number>/<protocol> in Dockerfile

### PR DESCRIPTION
**Pull request by coreydaley**
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1505558

### CODE REVIEW

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -1109,11 +1109,32 @@ func optionallyValidateExposedPorts(config *AppConfig, repositories app.SourceRe
...
 }
 
-func exposedPortsAreNumeric(node *dockerfileparser.Node) error {
+func exposedPortsAreValid(node *dockerfileparser.Node) error {
+	allErrs := make([]error, 0)
+
 	for _, port := range dockerfileutil.LastExposedPorts(node) {
-		if _, err := strconv.ParseInt(port, 10, 32); err != nil {
-			return fmt.Errorf("could not parse %q: must be numeric", port)
+		errs := make([]string, 0)
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Don't we already have parse function for PORT/PROTOCOL somewhere?

Comment (coreydaley) [reply pkg/generate/app/cmd/newapp.go]: Updated

**Comment by coreydaley** [comment]
Added test for updated exposedPortsAreValid function
@openshift/sig-developer-experience ptal

File: pkg/generate/app/cmd/newapp.go [hunk 2]
```diff
@@ -1109,11 +1110,38 @@ func optionallyValidateExposedPorts(config *AppConfig, repositories app.SourceRe
...
+
+		if strings.HasPrefix(port, "$") {
+			errs = append(errs, "args are not currently supported for port numbers, did you mean to use config.AllowNonNumericExposedPorts?")
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp.go]: i don't know what config.AllowNonNumericExposedPorts is or how to set it.  How will our users?

Comment (bparees) [reply pkg/generate/app/cmd/newapp.go]: (I don't think they even can. it's an internal setting we set to true when invoking new-build)

Comment (coreydaley) [reply pkg/generate/app/cmd/newapp.go]: updated

File: pkg/generate/app/cmd/newapp.go [hunk 3]
```diff
@@ -1109,11 +1110,38 @@ func optionallyValidateExposedPorts(config *AppConfig, repositories app.SourceRe
...
+func exposedPortsAreValid(node *dockerfileparser.Node) error {
+	allErrs := make([]error, 0)
+
 	for _, port := range dockerfileutil.LastExposedPorts(node) {
-		if _, err := strconv.ParseInt(port, 10, 32); err != nil {
-			return fmt.Errorf("could not parse %q: must be numeric", port)
+		errs := make([]string, 0)
+
+		if strings.HasPrefix(port, "$") {
+			errs = append(errs, "args are not currently supported for port numbers")
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp.go]: s/currently//

Comment (coreydaley) [reply pkg/generate/app/cmd/newapp.go]: corrected

**Comment by coreydaley** [comment]
@bparees removed "currently" from the test

**Comment by bparees** [comment]
@coreydaley this needs a followup.  If the port protocol is UDP then we need to setup the port on the container+service as such.  (Currently I think it's always getting set as TCP).

**Pull request merged**